File: 맞춰봐.py
def check(idx):
  sum=0
  for i in range(idx,-1,-1):
    sum+=ans[i]
    if sign[i][idx]==0:
      if sum !=0 : return False
    elif sign[i][idx] >0:
      if sum<=0: return False
    elif sign[i][idx] <0:
      if sum >=0  :  return False
  return True

def go(idx):
  if idx==n: return True
  if sign[idx][idx]==0:
    ans[idx]=0
    return check(idx) and go(idx+1)
  for i in range(1,11):
    ans[idx] = sign[idx][idx]*i
    if check(idx) and go(idx+1): return True
  return False

n=int(input())
s=input()
sign = [[0]*n for _ in range(n)]
ans=[0]*n
cnt=0
for i in range(n):
  for j in range(i,n):
    if s[cnt]=='0' : sign[i][j] = 0
    elif s[cnt]=='+' :sign[i][j] =1
    else: sign[i][j] = -1
    cnt+=1
go(0)
print(' '.join(map(str, ans)))

File: 맞춰봐.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <string>

using namespace std;
int sign[11][11];
int ans[11];
int n;

bool check(int index){
  int sum =0; 
  for(int i=index; i>=0; i--){
    sum+=ans[i];
    if(sign[i][index]==0){
      if(sum!=0) return false;
    }else if(sign[i][index] > 0){
      if(sum <= 0)return false;
    }else if(sign[i][index] <0){
      if(sum >=0) return false;
    }
  }
  return true;
}

bool go(int index){
  if(index==n) return true;
  if(sign[index][index]==0){
    ans[index]=0;
    return check(index)&&go(index+1);
  }
  for(int i=1; i<=10; i++){
    ans[index]= sign[index][index]*i;
    if(check(index)&&go(index+1)) return true;
  }
  return false;
}
int main(){
  string s;
  cin >> n ;
  cin>>s;
  int cnt =0;
  for(int i=0;i<n; i++){
    for(int j=i; j<n; j++){
      if(s[cnt]=='0') sign[i][j] = 0;
      else if (s[cnt]=='+') sign[i][j] = 1;
      else sign[i][j] = -1;
      cnt++;
    }
  }
  go(0);
  for(int i=0; i<n; i++){
    cout<<ans[i]<<' ';
  }
  cout<<'\n';
  return 0;
}